api/apps/api/src/modules/geo-features/geo-features.service.ts

```diff
@@ -138,19 +141,48 @@ export class GeoFeaturesService extends AppBaseService<
      * it is supplied as part of a GET query parsed according to the JSON:API
      * spec), and if a projectId is supplied in either way, we first check if
      * the project exists (if not, we throw a NotFoundException).
+     *
+     * In any case, when we have a projectId, we also limit the selection of
+     * features to those that intersect the project's bbox.
      */
 
     let queryFilteredByPublicOrProjectSpecificFeatures;
     const projectId: string | undefined =
       (info?.params?.projectId as string) ??
       (fetchSpecification?.filter?.projectId as string);
     if (projectId) {
-      await this.projectRepository.findOneOrFail(projectId).catch((_error) => {
-        throw new NotFoundException(`No project with id ${projectId} exists.`);
-      });
+      const project = await this.projectRepository
+        .findOneOrFail(projectId)
+        .then((project) => project)
+        .catch((_error) => {
+          throw new NotFoundException(
+            `No project with id ${projectId} exists.`,
+          );
+        });
+      const geoFeaturesWithinProjectBbox = await this.geoFeaturesGeometriesRepository
+        .createQueryBuilder('geoFeatureGeometries')
+        .where(
+          `st_intersects(
+        st_makeenvelope(:xmin, :ymin, :xmax, :ymax, 4326),
+        "geoFeatureGeometries".the_geom
+      )`,
+          {
+            xmin: project.bbox[0],
+            ymin: project.bbox[1],
+            xmax: project.bbox[2],
+            ymax: project.bbox[3],
+          },
+        )
+        .getMany()
+        .then((result) => result.map((i) => i.id))
+        .catch((error) => {
+          throw new Error(error);
+        });
+
       queryFilteredByPublicOrProjectSpecificFeatures = query.andWhere(
-        `${this.alias}.projectId = :projectId OR ${this.alias}.projectId IS NULL`,
-        { projectId },
+        `${this.alias}.projectId = :projectId OR ${this.alias}.projectId IS NULL
+        AND ${this.alias}.id IN (:...geoFeaturesWithinProjectBbox)`,
```

Comment:
@aagm I am not sure about this narrowing clause - see above where `geoFeaturesWithinProjectBbox` comes from and how it's computed (not sure about that query either). The intent is to limit our selection to features that somehow intersect the project planning area's bbox (when testing with a random Rwanda project I got ~100K rows out of 1.7M) which in turn should help with the sp-far-quite-expensive view that extracts properties (`geodb.feature_properties`).

Feedback welcome on any of this ❤️

Reply:
@aagm nevermind, this will never work with real data:

```
status: 500
title: "bind message has 5009 parameter formats but 0 parameters"
```

I have removed the `AND` part of the clause for the moment. I am thinking about how to deal with this spatial "join" across the two dbs without blowing things up at any scale - maybe we could set the bbox of each feature in the apidb at import time, and move the `st_intersects()` there (in a view?), but I am not too keen on this either...